src/main/java/net/minestom/server/inventory/InventoryImpl.java

```diff
@@ -0,0 +1,309 @@
+package net.minestom.server.inventory;
+
+import it.unimi.dsi.fastutil.ints.IntIterators;
+import net.minestom.server.entity.Player;
+import net.minestom.server.event.EventDispatcher;
+import net.minestom.server.event.inventory.InventoryItemChangeEvent;
+import net.minestom.server.inventory.click.*;
+import net.minestom.server.item.ItemStack;
+import net.minestom.server.network.packet.server.play.CloseWindowPacket;
+import net.minestom.server.network.packet.server.play.SetSlotPacket;
+import net.minestom.server.network.packet.server.play.WindowItemsPacket;
+import net.minestom.server.tag.TagHandler;
+import net.minestom.server.utils.MathUtils;
+import net.minestom.server.utils.validate.Check;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.function.UnaryOperator;
+
+sealed class InventoryImpl implements Inventory permits ContainerInventory, PlayerInventory {
+
+    private static final VarHandle ITEM_UPDATER = MethodHandles.arrayElementVarHandle(ItemStack[].class);
+
+    private final int size;
+    protected final ItemStack[] itemStacks;
+
+    protected final ClickPreprocessor clickPreprocessor = new ClickPreprocessor(this);
+
+    private final TagHandler tagHandler = TagHandler.newHandler();
+
+    // the players currently viewing this inventory
+    protected final Set<Player> viewers = new CopyOnWriteArraySet<>();
+    protected final Set<Player> unmodifiableViewers = Collections.unmodifiableSet(viewers);
+    // (player -> cursor item) map, used by the click listeners
+    protected final Map<Player, ItemStack> cursorPlayersItem = new ConcurrentHashMap<>();
+
+    public static final @NotNull ClickHandler DEFAULT_HANDLER = new StandardClickHandler(
+            (builder, item, slot) -> slot >= builder.clickedInventory().getSize() ?
+                    IntIterators.fromTo(0, builder.clickedInventory().getSize()) :
+                    PlayerInventory.getInnerShiftClickSlots(builder, item, slot),
+            (builder, item, slot) -> IntIterators.concat(
+                    IntIterators.fromTo(0, builder.clickedInventory().getSize()),
+                    PlayerInventory.getInnerDoubleClickSlots(builder, item, slot)
+            ));
+
+    protected InventoryImpl(int size) {
+        this.size = size;
+        this.itemStacks = new ItemStack[getSize()];
```

Comment:
This is the nitpick-iest of all nitpicks, but the `getSize` call seems weird here given that you have the size locally anyway